pkg/verifier/safe_delete.go

```diff
@@ -46,15 +53,56 @@ func SafeDelete(ctx context.Context, logger log.Logger, bkt objstore.Bucket, bac
 		}
 	}()
 
+	// Download the TSDB block.
+	dir := filepath.Join(tempdir, id.String())
 	if err := block.Download(ctx, logger, bkt, id, dir); err != nil {
 		return errors.Wrap(err, "download from source")
 	}
 
-	if err := block.Upload(ctx, logger, backupBkt, dir); err != nil {
+	// At this point call backupAndDeleteDownloaded() to handle the now local
+	// TSDB block.
+	return backupAndDeleteDownloaded(ctx, logger, dir, bkt, backupBkt, id)
+}
+
+// BackupAndDeleteDownloaded works much like BackupAndDelete in that it will
+// move a TSDB block from a bucket to a backup bucket. The bdir parameter
+// points to the location on disk where the TSDB block was previously
+// downloaded allowing this function to avoid downloading the TSDB block from
+// the source bucket again. An error is returned if any operation fails.
+func BackupAndDeleteDownloaded(ctx context.Context, logger log.Logger, bdir string, bkt, backupBkt objstore.Bucket, id ulid.ULID) error {
+	// Does this TSDB block exist in backupBkt already?
+	found, err := TSDBBlockExistsInBucket(ctx, backupBkt, id)
+	if err != nil {
+		return err
+	}
+	if found {
+		return errors.Errorf("%s dir seems to exists in backup bucket. Remove this block manually if you are sure it is safe to do", id)
+	}
+
+	// Now we call backupAndDeleteDownloaded() to handle the work.
+	return backupAndDeleteDownloaded(ctx, logger, bdir, bkt, backupBkt, id)
+}
+
+// backupAndDeleteDownloaded is a helper function that uploads a TSDB block
+// found on disk to the backup bucket and, on success, removes the TSDB block
+// from the source bucket.
+func backupAndDeleteDownloaded(ctx context.Context, logger log.Logger, bdir string, bkt, backupBkt objstore.Bucket, id ulid.ULID) error {
```

Comment:
Sorry for being picky but there is some improvement we can make here.

I like `BackupAndDelete` vs `BackupAndDeleteDownloaded`, but it's not clear what's the difference between `backupAndDeleteDownloaded`  and `BackupAndDeleteDownloaded` is. Sure they are heavily commented, but comment should have less priority than naming. I think here the proper naming and small possible signature of function would do the work.

So what about:
```suggestion
func backupDownloaded(ctx context.Context, logger log.Logger, bdir string, bktToRemove, backupBkt objstore.Bucket, id ulid.ULID) error {
```
and then if no error do on the caller in both `BackupAndDeleteDownloaded` and `BackupAndDelete` just.
```
// Block uploaded, so we are ok to remove from src bucket.
if err := block.Delete(ctx, bkt, id); err != nil {
	return errors.Wrap(err, "delete from source")
}
```
This will make the signature of  `BackupAndDeleteDownloaded`  and `backupDowloaded` smaller (thus reducing complexity) as you can drop `bkt` argument. 

What do you think? (:

Reply:
Adjusted as requested, and went through another round of testing.